README.md
## freeformatter

### A simple golang script that will do below:

    - [x] show all the MIME listings
    - [x] get a specific mime information based on the extention
    - [x] encode or decode a given string of data
    - [x] html escape a given string of data
    - [x] base64 encode or decode a given string of data

### Command


```sh

$ ./freeformatter

 Version:  0.1.0-0

  -b64decode string
        use to base64 encode/decode a data
  -b64decode-url string
        use to base64 encode/decode a data
  -b64encode string
        use to base64 encode/decode a data
  -b64encode-url string
        use to base64 encode/decode a data
  -decode string
        use to encode/decode a data
  -decode-url string
        use to encode/decode a data
  -encode string
        use to encode/decode a data
  -encode-url string
        use to encode/decode a data
  -html-esc string
        use to escape/unescape an HTML
  -html-esc-url string
        use to escape/unescape an HTML
  -mime-list
        use to show the list of MIME types
  -mime-type string
        use to show the list of MIME types
  -qr-code-gen string
        use to generate a PNG QR Code

-------------------------------------------------------


        ./freeformatter  --encode "test data here"

        ./freeformatter  --decode "test+data+here"

        ./freeformatter  --mime-type '.avi'

        ./freeformatter  --mime-list

        ./freeformatter  --qr-code-gen='{"data": "https://www.google.com.sg/","filename":"qrcode.png","size":256}'


-------------------------------------------------------

```



## Docker Binary

- [x] In order to  use it as dockerize binary


```sh

    sudo  sysctl -w net.ipv4.ip_forward=1

    sudo  docker run --rm  bayugyug/freeformatter 

    sudo  docker run --rm  bayugyug/freeformatter --mime-list

    sudo  docker run -v `pwd`:`pwd` -w `pwd` --rm  bayugyug/freeformatter         --qr-code-gen='{"data": "https://www.google.com.sg/","filename":"qrcode-hahaha.png","size":256}'

    sudo  docker run -v `pwd`:`pwd` -w `pwd` --rm  bayugyug/freeformatter:alpine  --qr-code-gen='{"data": "https://www.google.com.sg/","filename":"qrcode-hahaha.png","size":256}'

```


## As HTTP Server

- [x] In order to  use it via CURL/WGET or Browser


```sh

    sudo  sysctl -w net.ipv4.ip_forward=1

    sudo  docker run -p 7000-8000:7000-8000 -v `pwd`:`pwd` -w `pwd` -d --name freeformatter-alpine  bayugyug/freeformatter:alpine --http --port 7778

    curl -i -v 'http://192.168.2.121:7778/mime-type/?p=.mp4'

```



CHANGELOG.md
#CHANGELOG

###freeformatter





route.go
package main

import (
	"encoding/json"
	"fmt"
	"github.com/julienschmidt/httprouter"
	"io"
	"net/http"
	"os"
	"strconv"
	"strings"
)

func indexHandler(w http.ResponseWriter, r *http.Request, _ httprouter.Params) {

	msg := `

		http://localhost:7777/mime-type?p=123445

		http://localhost:7777/mime-list

		http://localhost:7777/enc-data?p=123445

		http://localhost:7777/enc-url?p=htp://www.google.com.sg/

		http://localhost:7777/dec-data?p=123445

		http://localhost:7777/dec-url?p=htp://www.google.com.sg/

		http://localhost:7777/b64-enc-data?p=123445

		http://localhost:7777/b64-enc-url?p=htp://www.google.com.sg/

		http://localhost:7777/b64-dec-datap=123445

		http://localhost:7777/b64-dec-url?p=htp://www.google.com.sg/

		http://localhost:7777/html-enc-data?p=123445

		http://localhost:7777/html-enc-url?p=htp://www.google.com.sg/

`
	fmt.Fprint(w, "Welcome to Freeformatter!\n", fmt.Sprintf("Version: %s\n\n\n%s", pVersion, msg))
}

func formatHandler(w http.ResponseWriter, r *http.Request, ps httprouter.Params) {
	//Content-Type: application/json
	var p = strings.TrimSpace(r.URL.Query().Get("p"))
	var m = strings.ToUpper(strings.TrimSpace(ps.ByName("mode")))

	//not-found
	v, ok := Formatters[m]
	if !ok {
		http.Error(w, http.StatusText(http.StatusUnauthorized), http.StatusUnauthorized)
		return
	}
	fmt.Println("RAW-DATA: ", p)
	rescode, result := v.Format(v.Mode, p)
	s := rescode
	if s == "" {
		s = "Success"
		//download-push-file
		if strings.EqualFold(m, "QR-CODE-GEN") {
			pushQrCode(w, r, pQRParams.Filename)
			return
		}
	}
	//good
	w.Header().Set("Content-Type", "application/json")
	jdata, _ := json.MarshalIndent(FormatterOutput{Result: result, Status: s}, "", "\t")
	fmt.Fprint(w, string(jdata))
}

func pushQrCode(w http.ResponseWriter, r *http.Request, filename string) {
	//Check if file exists and open
	fh, err := os.Open(filename)
	defer fh.Close() //Close after function return
	if err != nil {
		//File not found, send 404
		fmt.Println(err)
		http.Error(w, http.StatusText(http.StatusNotFound), http.StatusNotFound)
		return
	}
	//Get the Content-Type of the file
	fhdr := make([]byte, 512)
	//Copy the headers into the FileHeader buffer
	fh.Read(fhdr)
	//Get content type of file
	ftype := http.DetectContentType(fhdr)

	//Get the file size
	fstat, _ := fh.Stat()                        //Get info from file
	fsize := strconv.FormatInt(fstat.Size(), 10) //Get file size as a string

	//Send the headers
	w.Header().Set("Content-Disposition", "attachment; filename="+filename)
	w.Header().Set("Content-Type", ftype)
	w.Header().Set("Content-Length", fsize)

	//Send the file
	//We read 512 bytes from the file already so we reset the offset back to 0
	fh.Seek(0, 0)
	io.Copy(w, fh) //'Copy' the file to the client
}


handler.go
package main

import (
	"bufio"
	"crypto/tls"
	"encoding/base64"
	"encoding/json"
	"fmt"
	qrcode "github.com/skip2/go-qrcode"
	"html"
	"io/ioutil"
	"log"
	"net"
	"net/http"
	"net/url"
	//"path"
	"strings"
	"time"
)

func b(s string) *bufio.Reader { return bufio.NewReader(strings.NewReader(s)) }

//handler entry of the app
func handler() {

	//get mime type
	if pMimeType != "" {
		v := Formatters["MIME-TYPE"]
		_, resmsg := v.Format(v.Mode, pMimeType)
		fmt.Println(resmsg)
		return
	}
	//get mime list
	if pMimeList {
		v := Formatters["MIME-LIST"]
		rescode, resmsg := v.Format(v.Mode, pMimeType)
		showStatus(rescode, resmsg)
		return
	}
	//encode
	if pEncData != "" {
		v := Formatters["ENC-DATA"]
		rescode, resmsg := v.Format(v.Mode, pEncData)
		showStatus(rescode, resmsg)
		return
	}
	//encode-url
	if pEncUrl != "" {
		v := Formatters["ENC-URL"]
		rescode, resmsg := v.Format(v.Mode, pEncUrl)
		showStatus(rescode, resmsg)
		return
	}
	//decode
	if pDecData != "" {
		v := Formatters["DEC-DATA"]
		rescode, resmsg := v.Format(v.Mode, pDecData)
		showStatus(rescode, resmsg)
		return
	}
	//decode-url
	if pDecUrl != "" {
		v := Formatters["DEC-URL"]
		rescode, resmsg := v.Format(v.Mode, pDecUrl)
		showStatus(rescode, resmsg)
		return
	}
	//encode:b64
	if pBase64Enc != "" {
		v := Formatters["B64-ENC-DATA"]
		rescode, resmsg := v.Format(v.Mode, pBase64Enc)
		showStatus(rescode, resmsg)
		return
	}
	//encode-url:b64
	if pBase64UrlEnc != "" {
		v := Formatters["B64-ENC-URL"]
		rescode, resmsg := v.Format(v.Mode, pBase64UrlEnc)
		showStatus(rescode, resmsg)
		return
	}
	//decode:b64
	if pBase64Dec != "" {
		v := Formatters["B64-DEC-DATA"]
		rescode, resmsg := v.Format(v.Mode, pBase64Dec)
		showStatus(rescode, resmsg)
		return
	}
	//decode-url:b64
	if pBase64UrlDec != "" {
		v := Formatters["B64-DEC-URL"]
		rescode, resmsg := v.Format(v.Mode, pBase64UrlDec)
		showStatus(rescode, resmsg)
		return
	}
	//html-esc
	if pHtmlEsc != "" {
		v := Formatters["HTML-ENC-DATA"]
		rescode, resmsg := v.Format(v.Mode, pHtmlEsc)
		showStatus(rescode, resmsg)
		return
	}
	//html-esc-url
	if pHtmlUrlEsc != "" {
		v := Formatters["HTML-ENC-URL"]
		rescode, resmsg := v.Format(v.Mode, pHtmlUrlEsc)
		if !pHttpServe {
			showStatus(rescode, resmsg)
			return
		}
		return
	}

	//qrcode
	if pQRCodeGen != "" {
		v := Formatters["QR-CODE-GEN"]
		rescode, resmsg := v.Format(v.Mode, pQRCodeGen)
		if !pHttpServe {
			showStatus(rescode, resmsg)
			return
		}
		return
	}
	//serve http
	if pHttpServe {
		initHttpRouters()
		return
	}

}

func getMimes() []MIMETypes {
	var mimes []MIMETypes
	err := json.Unmarshal(jsonStr, &mimes)
	if err != nil {
		fmt.Println("error:", err)
	}
	return mimes
}

func fmtMime(mode, data string) (string, string) {
	mimelist := getMimes()
	var jdata []byte
	if strings.EqualFold(mode, "TYPE") {
		//type
		for _, v := range mimelist {
			if strings.EqualFold(v.Ext, data) {
				jdata, _ = json.MarshalIndent(v, "", "\t")
				break
			}
		}
	} else {
		//list
		jdata, _ = json.MarshalIndent(mimelist, "", "\t")
	}
	return "", string(jdata)
}

func fmtEnc(mode, data string) (string, string) {
	var s string
	if strings.EqualFold(mode, "URL") {
		if !strings.HasPrefix(data, "http") {
			return "Fail", "Invalid url parameter: " + data
		}
		_, s = getResult(data)
	} else {
		s = data
	}
	return "", url.QueryEscape(s)
}

func fmtDec(mode, data string) (string, string) {
	var s string
	if strings.EqualFold(mode, "URL") {
		if !strings.HasPrefix(data, "http") {
			return "Fail", "Invalid url parameter: " + data
		}

		_, s = getResult(data)
	} else {
		s = data
	}
	enc, _ := url.QueryUnescape(s)
	return "", enc
}

func fmtEncB64(mode, data string) (string, string) {
	var s string
	if strings.EqualFold(mode, "URL") {
		if !strings.HasPrefix(data, "http") {
			return "Fail", "Invalid url parameter: " + data
		}
		_, s = getResult(data)
	} else {
		s = data
	}
	enc := base64.StdEncoding.EncodeToString([]byte(s))
	return "", enc
}

func fmtDecB64(mode, data string) (string, string) {
	var s string

	if strings.EqualFold(mode, "URL") {
		if !strings.HasPrefix(data, "http") {
			return "Fail", "Invalid url parameter: " + data
		}
		_, s = getResult(data)
	} else {
		s = data
	}
	decoded, err := base64.StdEncoding.DecodeString(s)
	if err != nil {
		return "Fail", fmt.Sprintf("Decode error: %v", err)
	}
	return "", string(decoded)
}

func fmtHtmlEsc(mode, data string) (string, string) {
	var s string
	if strings.EqualFold(mode, "URL") {
		if !strings.HasPrefix(data, "http") {
			return "Fail", "Invalid url parameter: " + data
		}
		_, s = getResult(data)
	} else {
		s = data
	}
	return "", html.EscapeString(s)
}

func fmtHtmlUnEsc(mode, data string) (string, string) {
	var s string
	if strings.EqualFold(mode, "URL") {
		if !strings.HasPrefix(data, "http") {
			return "Fail", "Invalid url parameter: " + data
		}

		_, s = getResult(data)
	} else {
		s = data
	}
	return "", html.UnescapeString(s)
}

func fmtQRCode(mode, data string) (string, string) {

	//-qr-code-gen='{"data": "https://www.google.com.sg/","filename":"qrcode.png","size":256}'
	err := json.Unmarshal([]byte(data), &pQRParams)

	//sanity
	if err != nil {
		return "Fail", fmt.Sprintf("QR error: Invalid parameters! %v", err)
	}
	//sanity
	if pQRParams.Data == "" {
		return "Fail", "QR error: data is empty!"
	}
	var d, f string
	var s int

	//default
	s = pQRParams.Size
	if pQRParams.Size < 50 || pQRParams.Size > 512 {
		s = 128
	}
	//default
	d = strings.TrimSpace(pQRParams.Data)
	if len(d) > 4196 {
		d = d[:4196]
	}
	//default
	f = strings.TrimSpace(pQRParams.Filename)
	if f == "" {
		f = "qrcode.png"
	}
	dst := f //path.Join(pQRCodeTmp, f)
	//gen
	err = qrcode.WriteFile(d, qrcode.Medium, s, dst)
	if err != nil {
		return "Fail", fmt.Sprintf("QR error: failed to generate code! %v", err)
	}
	return "", dst
}

func showStatus(status, result string) {
	s := status
	if s == "" {
		s = "Success"
	}
	jdata, _ := json.MarshalIndent(FormatterOutput{Result: result, Status: s}, "", "\t")
	fmt.Println(string(jdata))
}

//getResult http req a url
func getResult(url string) (int, string) {
	//client
	c := &http.Client{Transport: &http.Transport{
		TLSClientConfig: &tls.Config{InsecureSkipVerify: true, RootCAs: pool},
		Dial: (&net.Dialer{
			Timeout: 300 * time.Second,
		}).Dial,
	},
	}
	res, err := c.Get(url)
	if err != nil {
		log.Println("ERROR: getResult:", err)
		return 0, ""
	}
	//get response
	robots, err := ioutil.ReadAll(res.Body)
	defer res.Body.Close()
	if err != nil {
		log.Println("ERROR: getResult:", err)
		return 0, ""
	}
	//give
	return res.StatusCode, string(robots)
}


global.go
package main

import (
	"crypto/x509"
	"flag"
	"fmt"
	"github.com/julienschmidt/httprouter"
	"io"
	"log"
	"math/rand"
	"net/http"
	"os"
	"path"
	"regexp"
	"strings"
	"time"
)

const (
	usageShowConsole  = "use to enable the output in console"
	usageMimeList     = "use to show the list of MIME types"
	usageDataFormat   = "use to encode/decode a data"
	usageBase64Format = "use to base64 encode/decode a data"
	usageHtmlFormat   = "use to escape/unescape an HTML"
	usageQRCode       = "use to generate a PNG QR Code"
	usageHttp         = "use to serve the results via HTTP"
	usageHttpPort     = "use to serve the results via HTTP @ Port No."
)

type Formatter struct {
	Mode   string
	Format func(mode string, data string) (string, string)
}

type FormatterOutput struct {
	Result string `json:"result,ommitempty`
	Status string `json:"status,ommitempty`
}

type QRCodeData struct {
	Data     string `json:"data,ommitempty`
	Filename string `json:"filename,ommitempty`
	Size     int    `json:"size,ommitempty`
}

var (
	pLogDir = "."
	//loggers
	infoLog  *log.Logger
	warnLog  *log.Logger
	errorLog *log.Logger

	//signal flag
	pStillRunning = true

	pBuildTime = "0"
	pVersion   = "0.1.0" + "-" + pBuildTime
	//console
	pShowConsole = true
	//envt
	pEnvVars = map[string]*string{
		"GMONGERS_LDIR": &pLogDir,
	}

	//print_format
	pPrintFormat  = "json"
	pHelp         = false
	pMimeList     = false
	pMimeType     = ""
	pEncData      = ""
	pDecData      = ""
	pEncUrl       = ""
	pDecUrl       = ""
	pBase64Enc    = ""
	pBase64Dec    = ""
	pBase64UrlEnc = ""
	pBase64UrlDec = ""
	pHtmlEsc      = ""
	pHtmlUrlEsc   = ""
	pQRCodeGen    = ""
	pHttpServe    = false

	//ssl certs
	pool *x509.CertPool

	Formatters map[string]Formatter

	pQRCodeTmp = "/tmp/freeformater-qrtmp"

	pHttpPort = "7777"
	pQRParams QRCodeData
)

func init() {
	//uniqueness
	rand.Seed(time.Now().UnixNano())
	//recovery
	initRecov()
	//evt
	initEnvParams()
	//loggers
	initLogger(os.Stdout, os.Stdout, os.Stderr)

	//init certs
	pool = x509.NewCertPool()
	pool.AppendCertsFromPEM(pemCerts)

	//in
	Formatters = map[string]Formatter{
		"MIME-TYPE":     {"TYPE", fmtMime},
		"MIME-LIST":     {"LIST", fmtMime},
		"ENC-DATA":      {"DATA", fmtEnc},
		"ENC-URL":       {"URL", fmtEnc},
		"DEC-DATA":      {"DATA", fmtDec},
		"DEC-URL":       {"URL", fmtDec},
		"B64-ENC-DATA":  {"DATA", fmtEncB64},
		"B64-ENC-URL":   {"URL", fmtEncB64},
		"B64-DEC-DATA":  {"DATA", fmtDecB64},
		"B64-DEC-URL":   {"URL", fmtDecB64},
		"HTML-ENC-DATA": {"DATA", fmtHtmlEsc},
		"HTML-ENC-URL":  {"URL", fmtHtmlEsc},
		"QR-CODE-GEN":   {"DATA", fmtQRCode},
	}

	//qr-tmf
	//initQRTmp()
}

//initHttpRouters, init the routing
func initHttpRouters() {
	pRouter := httprouter.New()
	pRouter.GET("/", indexHandler)
	pRouter.GET("/:mode", formatHandler)
	pRouter.POST("/:mode", formatHandler)
	pRouter.GET("/:mode/", formatHandler)
	pRouter.POST("/:mode/", formatHandler)
	fmt.Println("Freeformatter\n\nVer: "+pVersion+"\n\nReady to serve @ port:", strings.Replace(pHttpPort, ":", "", -1))
	log.Fatal(http.ListenAndServe(":"+pHttpPort, pRouter))
}

//initRecov is for dumpIng segv in
func initRecov() {
	//might help u
	defer func() {
		recvr := recover()
		if recvr != nil {
			fmt.Println("MAIN-RECOV-INIT: ", recvr)
		}
	}()
}

//os.Stdout, os.Stdout, os.Stderr
func initLogger(i, w, e io.Writer) {
	//just in case
	if !pShowConsole {
		infoLog = makeLogger(i, pLogDir, "freeformatter", "INFO: ")
		warnLog = makeLogger(w, pLogDir, "freeformatter", "WARN: ")
		errorLog = makeLogger(e, pLogDir, "freeformatter", "ERROR: ")
	} else {
		infoLog = log.New(i,
			"INFO: ",
			log.Ldate|log.Ltime|log.Lmicroseconds)
		warnLog = log.New(w,
			"WARN: ",
			log.Ldate|log.Ltime|log.Lshortfile)
		errorLog = log.New(e,
			"ERROR: ",
			log.Ldate|log.Ltime|log.Lshortfile)
	}
}

//initEnvParams enable all OS envt vars to reload internally
func initEnvParams() {
	//just in-case, over-write from ENV
	for k, v := range pEnvVars {
		if os.Getenv(k) != "" {
			*v = os.Getenv(k)
		}
	}
	//get options
	flag.BoolVar(&pMimeList, "mime-list", pMimeList, usageMimeList)
	flag.StringVar(&pMimeType, "mime-type", pMimeType, usageMimeList)

	flag.StringVar(&pEncData, "encode", pEncData, usageDataFormat)
	flag.StringVar(&pDecData, "decode", pDecData, usageDataFormat)

	flag.StringVar(&pEncUrl, "encode-url", pEncUrl, usageDataFormat)
	flag.StringVar(&pDecUrl, "decode-url", pDecUrl, usageDataFormat)

	flag.StringVar(&pBase64Enc, "b64encode", pBase64Enc, usageBase64Format)
	flag.StringVar(&pBase64Dec, "b64decode", pBase64Dec, usageBase64Format)

	flag.StringVar(&pBase64UrlEnc, "b64encode-url", pBase64UrlEnc, usageBase64Format)
	flag.StringVar(&pBase64UrlDec, "b64decode-url", pBase64UrlDec, usageBase64Format)

	flag.StringVar(&pHtmlEsc, "html-esc", pHtmlEsc, usageHtmlFormat)
	flag.StringVar(&pHtmlUrlEsc, "html-esc-url", pHtmlUrlEsc, usageHtmlFormat)

	flag.StringVar(&pQRCodeGen, "qr-code-gen", pQRCodeGen, usageQRCode)

	flag.BoolVar(&pHttpServe, "http", pHttpServe, usageHttp)

	flag.StringVar(&pHttpPort, "port", pHttpPort, usageHttpPort)
	flag.Parse()

	//either 1 should be present
	if pHelp {
		showUsage()
	}

	//either 1 should be present
	if !pMimeList &&
		pMimeType == "" &&
		pEncData == "" &&
		pDecData == "" &&
		pEncUrl == "" &&
		pDecUrl == "" &&
		pBase64Enc == "" &&
		pBase64Dec == "" &&
		pBase64UrlEnc == "" &&
		pBase64UrlDec == "" &&
		pHtmlEsc == "" &&
		pHtmlUrlEsc == "" &&
		!pHttpServe &&
		pQRCodeGen == "" {
		showUsage()
	}

}

//qrTmp try to init all filehandles for logs
func initQRTmp() {
	if _, err := os.Stat(pQRCodeTmp); os.IsNotExist(err) {
		//mkdir -p
		os.MkdirAll(pQRCodeTmp, os.ModePerm)
	}
}

//formatLogger try to init all filehandles for logs
func formatLogger(fdir, fname, pfx string) string {
	t := time.Now()
	r := regexp.MustCompile("[^a-zA-Z0-9]")
	p := t.Format("2006-01-02") + "-" + r.ReplaceAllString(strings.ToLower(pfx), "")
	s := path.Join(pLogDir, fdir)
	if _, err := os.Stat(s); os.IsNotExist(err) {
		//mkdir -p
		os.MkdirAll(s, os.ModePerm)
	}
	return path.Join(s, p+"-"+fname+".log")
}

//makeLogger initialize the logger either via file or console
func makeLogger(w io.Writer, ldir, fname, pfx string) *log.Logger {
	logFile := w
	if !pShowConsole {
		var err error
		logFile, err = os.OpenFile(formatLogger(ldir, fname, pfx), os.O_RDWR|os.O_CREATE|os.O_APPEND, 0664)
		if err != nil {
			log.Println(err)
		}
	}
	//give it
	return log.New(logFile,
		pfx,
		log.Ldate|log.Ltime|log.Lmicroseconds|log.Lshortfile)

}

//dumpW log into warning
func dumpW(s ...interface{}) {
	warnLog.Println(s...)
}

//dumpWF log into warning w/ fmt
func dumpWF(format string, s ...interface{}) {
	warnLog.Println(fmt.Sprintf(format, s...))
}

//dumpE log into error
func dumpE(s ...interface{}) {
	errorLog.Println(s...)
}

//dumpE log into error w/ fmt
func dumpEF(format string, s ...interface{}) {
	errorLog.Println(fmt.Sprintf(format, s...))
}

//dumpI log into info
func dumpI(s ...interface{}) {
	infoLog.Println(s...)
}

//dumpIF log into info
func dumpIF(format string, s ...interface{}) {
	infoLog.Println(fmt.Sprintf(format, s...))
}

//showUsage
func showUsage() {

	msg := `
-------------------------------------------------------


	./freeformatter  --encode "test data here"

	./freeformatter  --decode "test+data+here"

	./freeformatter  --mime-type '.avi'

	./freeformatter  --mime-list

	./freeformatter  --qr-code-gen='{"data": "https://www.google.com.sg/","filename":"qrcode.png","size":256}'

-------------------------------------------------------
`
	fmt.Println("Version: ", pVersion, "\n")
	flag.PrintDefaults()
	fmt.Println(msg)
	os.Exit(0)
}
